Make the JS AST -> source step of the linker incremental.

sjrd opened the pull request:
~~Based on #4815 for the purposes of performance analysis; the first two commits do not belong to this PR. But otherwise can be reviewed independently.~~

In `BasicLinkerBackend`, we now cache the JavaScript source code produced by each top-level tree, as well as the corresponding source map `Fragment`. Trees are identified with their reference identity.

The caches are stored per-`ModuleID`, since

* several modules can be written in parallel, and
* the emitter throws caches away across `ModuleID`s anyway.

This makes the tree-to-source step of our linker truly incremental.

The source maps building is incremental up to a `Fragment`. We cannot avoid a full pass for actually emitting the content, because the `name` table and `source` table are inherently global and cannot be incrementalized.

These changes bring about a 2x speedup to the "BasicBackend: Write result" step of the linker. This can be improved later by:

* caching more trees (some "rare, small" trees such as top-level exports, static initialization calls and some module-related trees
  are not cached in the `Emitter`), and
* caching the UTF-8-encoded bytes instead of the yet-to-be-encoded Strings.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/Emitter.scala:
```diff
@@ -695,6 +716,57 @@ final class Emitter(config: Emitter.Config) {
+      }
+
+      if (_tree == null || (_lastCtor ne ctor) || !allSame(_lastMemberMethods, memberMethods) ||
+          !allSame(_lastExportedMembers, exportedMembers)) {
```
I'm not sure where the right place to put this comment is:

I feel like this needs some smoke tests, because we start relying on reference equality. While I don't think this is a problem, it is a very non-local property, so IMO it's very easy we break caching without noticing it.

So I feel we need a test to make sure we actually get cache hits here (unclear to me how exactly this test would look though).

sjrd replied:
I added a new linker test `BasicLinkerBackendTest` that checks that a no-op second run does not invalidate anything (except the one module initializer tree). If the `eq`s here fail, that will be reflected in what the `BasicLinkerBackend` recomputes and therefore its stats logging.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/javascript/SourceMapWriter.scala:
```diff
@@ -73,10 +73,143 @@ private object SourceMapWriter {
+  }
+
+  final class FragmentBuilder extends Builder {
+    private val elements = new ListBuffer[FragmentElement]
```
Wouldn't an ArrayBuffer be better here?

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/BasicLinkerBackend.scala:
```diff
@@ -93,8 +107,141 @@ final class BasicLinkerBackend(config: LinkerBackendImpl.Config)
+    }
+  }
+
+  private final class PrintedTree(val jsCode: String, val sourceMapFragment: SourceMapWriter.Fragment) {
```
I can bring over the commit "Directly write bytes in the back-end instead of using a UTF-8 encoder." from #4817 over to this PR, if you prefer. I thought it would be easier to review it independently, but I can also bring it here.

gzm0 replied:
Actually, that commit addresses downstream concerns, but not what I meant here: It still prints the code to a string and then converts that to a byte array. I think we should do that processing in-stream.

Alternatively to keep it simple: Just add a TODO in the code here and we'll deal with it on the other PR.

sjrd replied:
I see. OK, I added a TODO comment for now, then.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/BasicLinkerBackend.scala:
```diff
@@ -93,8 +107,141 @@ final class BasicLinkerBackend(config: LinkerBackendImpl.Config)
+    }
+  }
+
+  private def extractLevelTrees(moduleBody: js.Tree): List[js.Tree] = {
```
I added a commit that changes what the emitter returns to `List`. It turns out it even had semantic value before, as demonstrated in the GCC back-end.

gzm0 replied:
CI says: The semantic required by GCC is different: It requires flattening of the inner blocks (at least that's my assumption).

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/BasicLinkerBackend.scala:
```diff
@@ -97,8 +107,132 @@ final class BasicLinkerBackend(config: LinkerBackendImpl.Config)
+
     logger.timeFuture("BasicBackend: Write result") {
       writer.write(moduleSet)
+    }.andThen { case result =>
```
`_ =>` ?

sjrd replied:
`_ =>` is not enough for it to be considered a `PartialFunction`, which `andThen` demands, but I made it `case _ =>`.

gzm0 commented on linker/shared/src/test/scala/org/scalajs/linker/BasicLinkerBackendTest.scala:
```diff
@@ -0,0 +1,84 @@
+      javalib <- TestIRRepo.javalib
+      allIRFiles = javalib ++ classDefsFiles
+      report1 <- linker.link(allIRFiles, initializers, outputDir, logger1)
+      report2 <- linker.link(allIRFiles, initializers, outputDir, logger2)
```
`report1` and `report2` seem unused. Use `_` instead?

gzm0 reviewed (changes requested):
Tree flattening seems problematic :-/

sjrd commented:
> Tree flattening seems problematic :-/

I fixed the flattening by manually flattening `js.Block`s in `Emitter.emitOnce`. I think we should push this flattening further inside the various generation functions, but that turns out to be much more involved. I have a WiP over at https://github.com/sjrd/scala-js/tree/backend-no-blocks-as-flattened-stats but it does not compile yet. We should take this as a future internal improvement.

---

Addressed comments and rebased onto `main`.

sjrd commented:
Rebased, for good measure.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/Emitter.scala:
```diff
@@ -267,19 +267,29 @@ final class Emitter(config: Emitter.Config) {
+         * them to gather several top-level trees under a single `js.Tree`.
+         * TODO We should improve this in the future.
+         */
+        val defTrees: List[js.Tree] = defTreesIterator.flatMap {
```
IIUC this will make the portions we cache on smaller (e.g. in ES 5.1, down to individual methods). Can you just confirm this is intended?

sjrd replied:
Yes and no. The basic linker backend would cache on the smaller level. But the emitter's `fullClass` caches at the class level, so it doesn't change anything.

(This was the same before the flattening changes you suggested, since the flattening was done by the `js.Block` anyway.)
